### PR TITLE
PLU-283: [INLINE-TILES-4] create new tile from dropdown

### PR DESCRIPTION
**Pull request by pregnantboy**
### TL;DR
This PR introduces the feature to create a new option via a modal to support creation of new tiles via the dropdown.

### What changed?
- A new component `AddNewOptionalModal` is added. This brings up the modal with an input field for the new option. The text in the modal is now hardcoded specifically for tile creation and can be extended in future. Also exposes a hook to call the mutation that corresponds to the `AddNewId` lives.
- Updated `ControlledAutocomplete` to render and manage the state for `AddNewOptionModal`. It handles the callbacks when the `create new option` is selected.
- Modified `SingleSelectProvider` to add 'Add New Option' button to the list of items.
- Moved `useIsFieldHidden` hook to `isFieldHidden` helper file
- Modified `SelectCombobox` and `ToggleChevron` to display `creating...` state

### How to test?
1. Create a pipe that uses Tiles -> Create Row action.
2. Optional: Open dev tools and throttle network speed to Slow 3G
3. In setup substep, select the `+ Create new Tile` option. This should bring up a modal to name the tile.
4. After submitting, you should see that the `Creating...` text and spinner.
5. The new tile should be created and submitted.
6. Also test that tabbing thru the input fields should not bring up the modal. (This was the previous incorrect behaviour for Add new connection)

### CODE REVIEW

**Comment by pregnantboy** [comment]
* **#663** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/663?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#661** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/661?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#660** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/660?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#659** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/659?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#657** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/657?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#648** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/648?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#647** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/647?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pregnantboy and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+      motionPreset="none"
+    >
+      <ModalOverlay />
+      <ModalContent>
+        <form onSubmit={onFormSubmit}>
+          <ModalHeader>{modalHeader}</ModalHeader>
+          <ModalCloseButton />
+          <ModalBody pb={8}>
+            <FormControl display="flex" flexDir="column" gap={2}>
+              {/* This is hardcoded for now, will change when there are more of such actions */}
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:102]: nit: ideally we should be able to reuse `CreateTileModal`. Maybe next person to hit this code should do the refactoring

File: packages/frontend/src/components/SingleSelect/types.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@ export type ComboboxItem<T = string> =
       icon?: As
       /** Badge to display on the right of the label, if available */
       badge?: JSX.Element
+      /** Whether this is an option to add new */
+      isAddNew?: boolean
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/SingleSelect/types.ts:18]: nit: wondering if it make sense to make this more generic, i.e. allow specifying onSelect callbacks on any item. so you have something like
* Extend `ComboBoxItem` to have a `callback` attribute
* define `hasOnSelectCallback` instead of `isItemAddNew` etc

But not a priority for now.

Comment (ogp-weeloong) [reply packages/frontend/src/components/SingleSelect/types.ts:18]: I just working through the `filteredString` stuff. Ok maybe a bit tricky to make this generic :/

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import { DropdownAddNewType } from '@plumber/types'
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx:1]: nit
```suggestion
import type { DropdownAddNewType } from '@plumber/types'
```

File: packages/frontend/src/components/ControlledAutocomplete/index.tsx [hunk 4]
```diff
@@ -76,55 +80,92 @@ function ControlledAutocomplete(
...
+    isOpen: isNewOptionModalOpen,
+  } = useDisclosure()
+
+  const { createNewOption, isCreatingNewOption } =
+    useCreateNewOption(fieldOnChange)
+
+  const onNewOptionModalSubmit = useCallback(
+    (inputValue: string) => {
+      onNewOptionModalClose()
+      createNewOption({ inputValue, addNewId: addNewOption?.id })
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/index.tsx:112]: thinking if better to close modal only if `createNewOption` succeeds (but that also means we need to update `createNewOption` to signal success / failure

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            break
+          }
+          default:
+            break
+        }
+        await client.refetchQueries({ include: [GET_DYNAMIC_DATA] })
+        if (newValue) {
+          setValue(newValue)
+        }
+      } finally {
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:59]: comment: stacey thinks we're relying too much on the toast error. ideally, we should show errors in the context of where the error occured (e.g. in this case - render as an error `Infobox` in the `Modal` and don't close it)

I think no need for now, but would be good to start moving the UI in that direction

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+} from '@chakra-ui/react'
+import { Button, FormLabel, Input } from '@opengovsg/design-system-react'
+import client from 'graphql/client'
+import { CREATE_TABLE } from 'graphql/mutations/tiles/create-table'
+import { GET_DYNAMIC_DATA } from 'graphql/queries/get-dynamic-data'
+
+interface AddNewOptionalModalProps {
+  modalHeader: string
+  onSubmit: (newValue: string) => void
+  onClose: () => void
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:20]: nit: `onSubmitError` callback would be nice

File: packages/frontend/src/components/ControlledAutocomplete/index.tsx [hunk 7]
```diff
@@ -76,55 +80,92 @@ function ControlledAutocomplete(
...
+                    isCreating: isCreatingNewOption,
+                  }
+                : undefined
+            }
+          />
+        </Box>
+      </Flex>
+      {isError && <FormErrorMessage>{error?.message}</FormErrorMessage>}
+      {addNewOption?.type === 'modal' && isNewOptionModalOpen && (
+        <AddNewOptionModal
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/index.tsx:162]: nit: probably can add comment to say we remount (instead of using the `isOpen` prop) because less dev effort to clear form data that way <.<. But not as performant

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx [hunk 8]
```diff
@@ -98,6 +109,22 @@ export const SingleSelectProvider = ({
 
   const { getItemByValue } = useLookupItems({ rawItems: allItems })
 
+  /**
+   * This create the + Add new option at the top
+   */
+  const allItemsWithAddNewOption = useMemo(() => {
+    if (addNew?.type === 'modal') {
+      const addNewByModalItem = {
+        value: '__ADD_NEW_PLACEHOLDER_VALUE__', // this is not referenced anywhere else
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx:118]: nit: would be nice to add this as a comment in `types.d.ts` e.g.

```
export interface IFieldDropdown extends IBaseField {
  type: 'dropdown'
  showOptionValue?: boolean
  allowArbitrary?: boolean
  // If this is enabled, `__ADD_NEW_PLACEHOLDER_VALUE__` becomes a reserved value.
  addNewOption?: {
    id: DropdownAddNewId // identifier when add new option is selected
    type: DropdownAddNewType
    label: string
  }
```

**Review by ogp-weeloong — APPROVED** [review]
Generally LGTM. Some small asks about providing error callbacks / signal success or failure in create new modal, but not a priority unless we got time.

Tested:
- Create new modal works
- Searching for somethign stops the "Create new" item from appearing
- "Create new" item is not selected by the singleselected when clicked
- Modal form is cleared when user closes and re-selects create new
- "Create new" item does not show up if freesolo is true and something is in the input

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx (hunk 1 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:102]: out of scope for now. this modal was meant to be a generic one input modal for other add-new dropdowns in future .

File: packages/frontend/src/components/SingleSelect/types.ts (hunk 2 above)
Comment (pregnantboy) [reply packages/frontend/src/components/SingleSelect/types.ts:18]: true that but isAddNew items do not set the value. whereas on select callbacks can just depend on a useEffect when the value is changed

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx (hunk 3 above)
Comment (pregnantboy) [reply packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx:1]: :+1:

File: packages/frontend/src/components/ControlledAutocomplete/index.tsx (hunk 4 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/index.tsx:112]: this is to re-use the create new option behaviour for inline and modal type

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx (hunk 5 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:59]: 👍🏻

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx (hunk 6 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx:20]: wont fail now since it doesnt call any mutation

File: packages/frontend/src/components/ControlledAutocomplete/index.tsx (hunk 7 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/index.tsx:162]: true

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx (hunk 8 above)
Comment (pregnantboy) [reply packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx:118]: added in SingleSelect/types.ts

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx (hunk 3 above)
Comment (pregnantboy) [reply packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx:1]: this was added in the later PR

**Comment by pregnantboy** [comment]
### Merge activity

* **Jul 4, 3:34 AM EDT**: @pregnantboy started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/659).
* **Jul 4, 3:39 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/659) rebased this pull request as part of a merge.
* **Jul 4, 3:40 AM EDT**: @pregnantboy merged this pull request with [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/659).
